Copy method for model and constraints

IgnaceBleukx opened the pull request.

tias commented on cpmpy/expressions/core.py:
```diff
@@ -131,6 +134,23 @@ def is_bool(self):
     def value(self):
         return None # default
 
+    def _copy_args(self, memodict={}):
```
add a one line comment what it does, and also why it exists (versus just being implemented in copy())

tias commented on cpmpy/expressions/core.py:
```diff
@@ -131,6 +134,23 @@ def is_bool(self):
+            copied.append(memodict[arg])
+        return copied
+
+    def copy(self, memodict = {}):
```
Add docstring

tias commented on cpmpy/expressions/core.py:
```diff
@@ -315,6 +339,11 @@ def value(self):
         elif self.name == ">=": return (arg_vals[0] >= arg_vals[1])
         return None # default
 
+    def copy(self, memodict={}):
```
Add docstring

tias commented on cpmpy/expressions/globalconstraints.py:
```diff
@@ -138,6 +138,10 @@ def decompose(self):
         """
         return None
 
+    def copy(self, memodict={}):
```
Add docstring... same for all the rest : )

tias commented on cpmpy/model.py:
```diff
@@ -231,4 +231,18 @@ def from_file(fname):
+
+    def copy(self, memodict={}):
+        """
+            Copies a the model to a new instance.
```
Returns a copy of the model?

something about the 'memodict' argument too.

and something about it being a shallow copy or deep copy?

IgnaceBleukx replied:
I was thinking about shallow/deep copy distinctions as well. But I can see no reason why someone would need a shallow copy of a model.
As variables are still shared between models then, solving one model will affect the variables in the other one.
Might still be usefull to clarify that it is a deep copy anyway indeed

IgnaceBleukx commented:
Updated the pullrequest with results of our earlier discussion.
Renamed the newly implemented copy methods to `deepcopy` and added a shallow copy method for models.
